components/eamxx/src/physics/dp/impl/dp_advance_iop_nudging_impl.hpp

```diff
@@ -13,10 +13,59 @@ namespace dp {
 
 template<typename S, typename D>
 KOKKOS_FUNCTION
-void Functions<S,D>::advance_iop_nudging(const Int& plev, const Spack& scm_dt, const Spack& ps_in, const uview_1d<const Spack>& t_in, const uview_1d<const Spack>& q_in, const uview_1d<Spack>& t_update, const uview_1d<Spack>& q_update, const uview_1d<Spack>& relaxt, const uview_1d<Spack>& relaxq)
+void Functions<S,D>::advance_iop_nudging(
+    // Input arguments
+    const Int& plev,
+    const Scalar& scm_dt,
+    const Scalar& ps_in,
+    const uview_1d<const Spack>& t_in,
+    const uview_1d<const Spack>& q_in,
+    const uview_1d<const Spack>& hyai,
+    const uview_1d<const Spack>& hyam,
+    const uview_1d<const Spack>& hybi,
+    const uview_1d<const Spack>& hybm,
+    // Kokkos stuff
+    const MemberType& team,
+    const Workspace& workspace,
+    // Output arguments
+    const uview_1d<Spack>& t_update,
+    const uview_1d<Spack>& q_update,
+    const uview_1d<Spack>& relaxt,
+    const uview_1d<Spack>& relaxq)
 {
-  // TODO
-  // Note, argument types may need tweaking. Generator is not always able to tell what needs to be packed
+  // Local variables
+  uview_1d<Spack>
+    pmidm1, // pressure at model levels
+    pintm1, // pressure at model interfaces (dim=plev+1)
+    pdelm1, // pdel(k)   = pint  (k+1)-pint  (k)
+    rtau;
+  workspace.template take_many_contiguous_unsafe<4>(
+    {"pmidm1", "pintm1", "pdelm1", "rtau"},
+    {&pmidm1, &pintm1, &pdelm1, &rtau});
+
+  // Get vertical level profiles
+  plevs0(plev, ps_in, hyai, hyam, hybi, hybm, team, pintm1, pmidm1, pdelm1);
+
+  const Int plev_pack = ekat::npack<Spack>(plev);
+  constexpr Scalar iop_nudge_tq_low = DPC::iop_nudge_tq_low;
+  constexpr Scalar iop_nudge_tq_high = DPC::iop_nudge_tq_high;
+  constexpr Scalar iop_nudge_tscale = DPC::iop_nudge_tscale;
+
+
+  Kokkos::parallel_for(Kokkos::TeamVectorRange(team, plev_pack), [&] (Int k) {
+    relaxt(k) = 0;
+    relaxq(k) = 0;
+
+    const auto condition = pmidm1(k) <= iop_nudge_tq_low*100
+                           &&
+                           pmidm1(k) >= iop_nudge_tq_high*100;
+    rtau(k).set(condition, std::max(scm_dt, iop_nudge_tscale));
+    relaxt(k).set(condition, (t_in(k) /*- tobs(k)*/)/rtau(k));
+    relaxq(k).set(condition, (q_in(k) /*- qobs(k)*/)/rtau(k));
+
+    t_update(k).set(condition, t_in(k) + relaxt(k)*scm_dt);
+    q_update(k).set(condition, q_in(k) + relaxq(k)*scm_dt);
```

Comment:
@bogensch 2 questions:
1. In lines 63, 64 and 66, 67, I use `t_in` and `q_in` but in the .F90 `t_update` and `q_update` are used. In F90 these `*_update` are `intent(out)` variables, and the `*_in` variables are never used which makes me think it was a typo. Either way, where the function is called, the same variable was used for both inputs, so it wouldn't change the answer. 
2. What is `tobs` and `qobs`?

Reply:
@tcclevenger , I'm having to deal with `tobs` and `qobs` in my current function as well:

```
    const uview_1d<Spack>& tobs,       // actual temperature, dims=(plev)                                                                                                                                                      
    const uview_1d<Spack>& qobs);      // actual W.V. Mixing ratio, dims=(plev) 
```

Reply:
Sorry for the late reply, for some reason just seeing this.  

1.  Whoops, yes that is a typo in the F90 but yes wouldn't change the answer.
2. tobs = observed temperature [K] ; qobs = observed water vapor [kg/kg]